examples/todo-list/src/repositories/todo.repository.ts

```diff
@@ -35,4 +37,46 @@ export class TodoRepository extends DefaultCrudRepository<
       todoListRepositoryGetter,
     );
   }
+
+  async find(
+    filter?: Filter<Todo>,
+    options?: Options,
+  ): Promise<TodoWithRelations[]> {
+    // Prevent juggler for applying "include" filter
+    // Juggler is not aware of LB4 relations
+    const include = filter && filter.include;
+    filter = filter && Object.assign(filter, {include: undefined});
+
+    const result = await super.find(filter, options);
+
+    // this is a temporary implementation, please see https://github.com/strongloop/loopback-next/issues/3195
+    if (include && include.length && include[0].relation === 'todoList') {
+      await Promise.all(
+        result.map(async r => {
+          r.todoList = await this.todoList(r.id);
+        }),
+      );
+    }
+
+    return result;
+  }
+
+  async findById(
+    id: typeof Todo.prototype.id,
+    filter?: Filter<Todo>,
+    options?: Options,
+  ): Promise<TodoWithRelations> {
+    // Prevent juggler for applying "include" filter
+    // Juggler is not aware of LB4 relations
```

Comment:
Since juggler is not aware of LB4 relations, perhaps we should move the code deleting `filter.include` to the base repository class `DefaultCrudRepository`, so that we don't have to repeat this logic everywhere.

What do you think?

/cc @raymondfeng @strongloop/loopback-maintainers

Reply:
The possible downside I see in such move:

Existing application don't have code to handle `filter.include`, it's passed down to juggler which rejects it and as a result, the requests end with an error (hopefully 400 Bad Request, possibly 500 Internal Server Error). The client knows that the `include` filter property is not supported.

If we change `DefaultCrudRepository` to discard `include`, then juggler will happily run the query and return back results, but with no related models. From the client's perspective, the `include` argument seems to be silently ignored :( I consider that a poor user experience.

I am proposing the following solution:

Modify `DefaultCrudRepository` to understand `filter.include`, call a new method that custom Repository implementations can override to supply their own inclusion resolution, and implement that new method by throwing a helpful error.

Implementation in `DefaultCrudRepository`:

```ts
{
  async find(
    filter?: Filter<T>,
    options?: Options,
  ): Promise<(T & Relations)[]> {
    const jugglerFilter = {...filter, include: undefined} as legacy.Filter;
    const models = await ensurePromise(
      this.modelClass.find(jugglerFilter, options),
    );
    const entities = this.toEntities(models);
    await this._includeRelatedModels(entities, filter, options);
    return entities;
  }

  // also modify findOne and findById similarly

  protected async _includeRelatedModels(
    entities: (T & Relations)[],
    filter: Filter<T>,
    options?: Options,
  ): Promise<void> {
    const msg = 'Inclusion of related models is not supported yet. ' + 
      'Please remove "include" property from the "filter" parameter.';
    const err = new Error(msg);
    err.code = 'FILTER_INCLUDE_NOT_SUPPORTED';
    throw err;
  }
}
```

With this change in place, the repositories in our example-todo-list application can supply custom implementation of `_includeRelatedModels` and don't have to overwrite all `find*` methods.

The missing piece is mapping `FILTER_INCLUDE_NOT_SUPPORTED` error code to HTTP status code, this should be done in `reject.provider.ts`:

https://github.com/strongloop/loopback-next/blob/23f91c19844d803347fa517fb2c4d66391692148/packages/rest/src/providers/reject.provider.ts#L14-L16

Reply:
On the second thoughts, maybe the duplication is ok in this stage, because the duplicated code is only temporary and will be removed when we figure out the inclusion resolver? By keeping the  changes in example-todo-list only and not modifying DefaultCrudRepository, we will keep our options open. If we realize that we need a different API for `_includeRelatedModels` or even a different design, then we have free hands to implement anything we like.

Thoughts?

Reply:
I actually like the proposal to add `_includeRelatedModels` in `DefaultCrudRepository` and require custom repositories to supply their own version of this function.

Reply:
I think we should wait until https://github.com/strongloop/loopback-next/issues/3195 is done especially so +1 to waiting in case we want a different approach.